### PR TITLE
PlaylistPage: fix search input

### LINKED ISSUE #38786: PlaylistPage: Unable to search through playlists
**What happened**:
I tried searching through the playlists but couldn't.
**What you expected to happen**:
I should be able to search through the playlist
**How to reproduce it (as minimally and precisely as possible)**:
On the sidemenu, hover/click on the dashboard icon
Navigate to the playlist
Try searching on the searchbar and notice it doesn't show any result.
**Anything else we need to know?**:

**Environment**:
- Grafana version: 8.1.0
- Data source type & version:
- OS Grafana is installed on: MacOS
- User OS & Browser: Chrome
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by thisisobate**
**What this PR does / why we need it**:
In Grafana versions < 7.5, we did not support search in playlist page. With that being said, we introduced search in 8.0 but unfortunately, the search API does not support wildcard search which makes result filtering impossible. This PR fixes this.
**Which issue(s) this PR fixes**:

Fixes #38786

### CODE REVIEW

**Review by torkelo — CHANGES_REQUESTED** [review]
This seems like a bad way to fix a bug :) 

Instead make wildcard search work:

```diff
diff --git a/pkg/services/sqlstore/playlist.go b/pkg/services/sqlstore/playlist.go
index 9abfc9f3eb..253679d25d 100644
--- a/pkg/services/sqlstore/playlist.go
+++ b/pkg/services/sqlstore/playlist.go
@@ -133,7 +133,7 @@ func SearchPlaylists(query *models.GetPlaylistsQuery) error {
        sess := x.Limit(query.Limit)

        if query.Name != "" {
-               sess.Where("name LIKE ?", query.Name)
+               sess.Where("name LIKE ?", "%" + query.Name + "%")
        }

        sess.Where("org_id = ?", query.OrgId)
```

```diff
diff --git a/public/app/features/playlist/PlaylistPage.tsx b/public/app/features/playlist/PlaylistPage.tsx
index 69a11f8321..b535e5ecf7 100644
--- a/public/app/features/playlist/PlaylistPage.tsx
+++ b/public/app/features/playlist/PlaylistPage.tsx
@@ -28,7 +28,8 @@ export const PlaylistPage: FC<PlaylistPageProps> = ({ navModel }) => {

   const { value: playlists, loading } = useAsync(async () => {
     return getAllPlaylist(searchQuery) as Promise<PlaylistDTO[]>;
-  }, [forcePlaylistsFetch]);
+  }, [forcePlaylistsFetch, searchQuery]);
+
   const hasPlaylists = playlists && playlists.length > 0;
   const onDismissDelete = () => setPlaylistToDelete(undefined);
   const onDeletePlaylist = () => {
```

With this search works, only some bug with re-mounts that cause the search input to lose focus. Any idea why that is?

File: public/app/features/playlist/PlaylistPage.tsx [hunk 1]
```diff
@@ -87,13 +84,6 @@ export const PlaylistPage: FC<PlaylistPageProps> = ({ navModel }) => {
   return (
     <Page navModel={navModel}>
       <Page.Contents isLoading={loading}>
-        {hasPlaylists && (
-          <PageActionBar
-            searchQuery={searchQuery}
-            linkButton={{ title: 'New playlist', href: '/playlists/new' }}
-            setSearchQuery={setSearchQuery}
-          />
-        )}
```
Comment (torkelo) [inline public/app/features/playlist/PlaylistPage.tsx:96]: with this you also removed the button to create new playlists :)

Comment (thisisobate) [reply public/app/features/playlist/PlaylistPage.tsx:96]: ah...I see.
Great catch!

**Comment by hugohaggmark** [comment]
> This seems like a bad way to fix a bug :)
> 
> Instead make wildcard search work:
> 
> ```diff
> diff --git a/pkg/services/sqlstore/playlist.go b/pkg/services/sqlstore/playlist.go
> index 9abfc9f3eb..253679d25d 100644
> --- a/pkg/services/sqlstore/playlist.go
> +++ b/pkg/services/sqlstore/playlist.go
> @@ -133,7 +133,7 @@ func SearchPlaylists(query *models.GetPlaylistsQuery) error {
>         sess := x.Limit(query.Limit)
> 
>         if query.Name != "" {
> -               sess.Where("name LIKE ?", query.Name)
> +               sess.Where("name LIKE ?", "%" + query.Name + "%")
>         }
> 
>         sess.Where("org_id = ?", query.OrgId)
> ```
> 
> ```diff
> diff --git a/public/app/features/playlist/PlaylistPage.tsx b/public/app/features/playlist/PlaylistPage.tsx
> index 69a11f8321..b535e5ecf7 100644
> --- a/public/app/features/playlist/PlaylistPage.tsx
> +++ b/public/app/features/playlist/PlaylistPage.tsx
> @@ -28,7 +28,8 @@ export const PlaylistPage: FC<PlaylistPageProps> = ({ navModel }) => {
> 
>    const { value: playlists, loading } = useAsync(async () => {
>      return getAllPlaylist(searchQuery) as Promise<PlaylistDTO[]>;
> -  }, [forcePlaylistsFetch]);
> +  }, [forcePlaylistsFetch, searchQuery]);
> +
>    const hasPlaylists = playlists && playlists.length > 0;
>    const onDismissDelete = () => setPlaylistToDelete(undefined);
>    const onDeletePlaylist = () => {
> ```
> 
> With this search works, only some bug with re-mounts that cause the search input to lose focus. Any idea why that is?

I feel I need to clarify my decision to skip the search here a bit. There was no search on this page until the Angular to React migration when I messed up in that migration and missed the search. So instead of extending the bug fix with backend code, I discussed this with Uche and told him he could remove search and add a feature request issue to improve the backend api. When that api improvement in place, we could add back the search again.

**Comment by torkelo** [comment]
The backend already has a query param that that is used to set the query.Name prop, was just missing this part:
sess.Where("name LIKE ?", "%" + query.Name + "%")

**Comment by thisisobate** [comment]
> With this search works, only some bug with re-mounts that cause the search input to lose focus. Any idea why that is?

Hmm, noticed that too. Would investigate this as well

File: public/app/features/playlist/EmptyQueryListBanner.tsx [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import { useStyles } from '@grafana/ui';
+import { GrafanaTheme } from '@grafana/data';
+import { css } from '@emotion/css';
+
+export const EmptyQueryListBanner = () => {
+  const styles = useStyles(getStyles);
+  return <div className={styles.noResult}>No playlist found!</div>;
+};
+
+const getStyles = (theme: GrafanaTheme) => {
```
Comment (torkelo) [inline public/app/features/playlist/EmptyQueryListBanner.tsx:11]: please us the new theme

File: public/app/features/playlist/PlaylistPage.tsx [hunk 3]
```diff
@@ -1,34 +1,43 @@
...
   const [startPlaylist, setStartPlaylist] = useState<PlaylistDTO | undefined>();
   const [playlistToDelete, setPlaylistToDelete] = useState<PlaylistDTO | undefined>();
   const [forcePlaylistsFetch, setForcePlaylistsFetch] = useState(0);
 
   const { value: playlists, loading } = useAsync(async () => {
-    return getAllPlaylist(searchQuery) as Promise<PlaylistDTO[]>;
-  }, [forcePlaylistsFetch]);
+    return getAllPlaylist(searchQuery);
+  }, [forcePlaylistsFetch, debouncedSearchQuery]);
+  useEffect(() => {
```
Comment (torkelo) [inline public/app/features/playlist/PlaylistPage.tsx:34]: Nitpick: would place new line above this useEffect to make it easier to read the code.

**Review by torkelo — APPROVED** [review]
Looks great, awesome job on fixing the tricky state issues! 

There is one minor issue when you search and get no results, then click on clear (to remove search query), the empty list banner shows for a second before the full list is loaded

**Pull request merged**